File: frontend/src/lib/humanInput.ts
/** DeerFlow-native human_input protocol (align with upstream Human Input Card). */

export type HumanInputMode = "free_text" | "single_choice" | "choice_with_other" | "form";

export type HumanInputOption = {
  id: string;
  label: string;
  value: string;
};

export type HumanInputFieldType =
  | "text"
  | "textarea"
  | "number"
  | "select"
  | "multi_select"
  | "checkbox"
  | "date";

export type HumanInputField = {
  name: string;
  label: string;
  type: HumanInputFieldType;
  required: boolean;
  placeholder?: string;
  options?: HumanInputOption[];
};

export type HumanInputRequest = {
  version: 1 | 2;
  kind: "human_input_request";
  source: string;
  request_id: string;
  tool_call_id?: string;
  clarification_type?: string;
  title?: string;
  question: string;
  context?: string | null;
  input_mode: HumanInputMode;
  options?: HumanInputOption[];
  fields?: HumanInputField[];
  call_id?: string;
};

export type HumanInputResponse =
  | {
      version: 1;
      kind: "human_input_response";
      source: string;
      request_id: string;
      response_kind: "option";
      option_id: string;
      value: string;
    }
  | {
      version: 1;
      kind: "human_input_response";
      source: string;
      request_id: string;
      response_kind: "text";
      value: string;
    };

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function nonEmpty(value: unknown): value is string {
  return typeof value === "string" && value.trim().length > 0;
}

function parseOptions(value: unknown): HumanInputOption[] | undefined {
  if (value === undefined) return undefined;
  if (!Array.isArray(value)) return undefined;
  const options: HumanInputOption[] = [];
  const seen = new Set<string>();
  for (const [idx, item] of value.entries()) {
    if (typeof item === "string" && item.trim()) {
      const id = `opt-${idx + 1}`;
      options.push({ id, label: item.trim(), value: item.trim() });
      continue;
    }
    if (!isRecord(item)) return undefined;
    const id = nonEmpty(item.id) ? item.id : `opt-${idx + 1}`;
    const label = nonEmpty(item.label) ? item.label : nonEmpty(item.value) ? item.value : null;
    const optionValue = nonEmpty(item.value) ? item.value : label;
    if (!label || !optionValue || seen.has(id)) return undefined;
    seen.add(id);
    options.push({ id, label, value: optionValue });
  }
  return options;
}

/** Harness often embeds "1. … / 2. …" inside question markdown — lift them into options. */
export function extractOptionsFromMarkdown(text: string): HumanInputOption[] {
  const options: HumanInputOption[] = [];
  const seen = new Set<string>();
  const re = /^\s*(?:\d+[\.\、\)]\s+|[-*]\s+)(.+?)\s*$/gm;
  let match: RegExpExecArray | null;
  let idx = 0;
  while ((match = re.exec(text)) !== null) {
    const label = match[1].trim();
    if (!label || seen.has(label)) continue;
    seen.add(label);
    idx += 1;
    options.push({ id: `opt-${idx}`, label, value: label });
  }
  return options;
}

export function splitClarificationMarkdown(text: string): {
  context: string | null;
  question: string;
  options: HumanInputOption[];
} {
  const raw = text.trim();
  if (!raw) {
    return { context: null, question: "需要你补充一些信息才能继续", options: [] };
  }
  const options = extractOptionsFromMarkdown(raw);
  let body = raw;
  if (options.length) {
    const lines = raw.split(/\r?\n/);
    const kept: string[] = [];
    for (const line of lines) {
      if (/^\s*(?:\d+[\.\、\)]\s+|[-*]\s+)/.test(line)) break;
      kept.push(line);
    }
    body = kept.join("\n").trim();
  }
  body = body.replace(/^[🧩❓ℹ️]\s*/, "").trim();
  let context: string | null = null;
  let question = body || raw;
  const qMatch = body.match(/第\s*\d+\s*个问题[:：]\s*/);
  if (qMatch && qMatch.index !== undefined) {
    const before = body.slice(0, qMatch.index).trim();
    const after = body.slice(qMatch.index + qMatch[0].length).trim();
    if (before) context = before;
    if (after) question = after;
  } else {
    const parts = body.split(/\n\s*\n/).map((p) => p.trim()).filter(Boolean);
    if (parts.length >= 2) {
      context = parts[0];
      question = parts.slice(1).join("\n\n");
    }
  }
  question = question.replace(/^第\s*\d+\s*个问题[:：]\s*/, "").trim() || question;
  return { context, question, options };
}

function parseFields(value: unknown): HumanInputField[] | undefined {
  if (value === undefined) return undefined;
  if (!Array.isArray(value)) return undefined;
  const fields: HumanInputField[] = [];
  const seen = new Set<string>();
  for (const field of value) {
    if (!isRecord(field) || !nonEmpty(field.name) || seen.has(field.name)) return undefined;
    seen.add(field.name);
    const label = nonEmpty(field.label) ? field.label : field.name;
    const type = (nonEmpty(field.type) ? field.type : "text") as HumanInputFieldType;
    const options = parseOptions(field.options);
    if (field.options !== undefined && options === undefined) return undefined;
    fields.push({
      name: field.name,
      label,
      type,
      required: field.required === true,
      ...(nonEmpty(field.placeholder) ? { placeholder: field.placeholder } : {}),
      ...(options ? { options } : {}),
    });
  }
  return fields;
}

export function parseHumanInputRequest(value: unknown): HumanInputRequest | null {
  if (!isRecord(value)) return null;
  let question = nonEmpty(value.question) ? value.question.trim() : null;
  if (!question) return null;

  const callId = nonEmpty(value.tool_call_id)
    ? value.tool_call_id
    : nonEmpty(value.call_id)
      ? value.call_id
      : undefined;
  const requestId = nonEmpty(value.request_id)
    ? value.request_id
    : callId
      ? `clarification:${callId}`
      : `clarification:${question.slice(0, 24)}`;

  let inputMode = value.input_mode;
  let options = parseOptions(value.options);
  let context = value.context;
  // Legacy / harness markdown blobs: lift numbered options out of the question body.
  if ((!options || options.length === 0) && (/\n\s*\d+[\.\、\)]/.test(question) || /第\s*\d+\s*个问题/.test(question))) {
    const split = splitClarificationMarkdown(question);
    question = split.question;
    options = split.options.length ? split.options : options;
    if ((context === undefined || context === null || context === "") && split.context) {
      context = split.context;
    }
  }
  const fields = parseFields(value.fields);
  if (inputMode !== "free_text" && inputMode !== "single_choice" && inputMode !== "choice_with_other" && inputMode !== "form") {
    inputMode = fields && fields.length > 0
      ? "form"
      : options && options.length > 0
        ? "choice_with_other"
        : "free_text";
  }
  if (
    (inputMode === "single_choice" || inputMode === "choice_with_other")
    && (!options || options.length === 0)
  ) {
    inputMode = "free_text";
  }
  if (inputMode === "form" && (!fields || fields.length === 0)) {
    inputMode = options && options.length > 0 ? "choice_with_other" : "free_text";
  }

  return {
    version: value.version === 2 || inputMode === "form" ? 2 : 1,
    kind: "human_input_request",
    source: nonEmpty(value.source) ? value.source : "ask_clarification",
    request_id: requestId,
    ...(callId ? { tool_call_id: callId, call_id: callId } : {}),
    ...(nonEmpty(value.clarification_type) ? { clarification_type: value.clarification_type } : {}),
    ...(nonEmpty(value.title) ? { title: value.title } : {}),
    question,
    ...(context === undefined || context === null || typeof context === "string"
      ? { context: typeof context === "string" ? context : null }
      : {}),
    input_mode: inputMode as HumanInputMode,
    ...(options && options.length ? { options } : {}),
    ...(fields ? { fields } : {}),
  };
}

export function createHumanInputOptionResponse(
  request: HumanInputRequest,
  option: HumanInputOption,
): HumanInputResponse {
  return {
    version: 1,
    kind: "human_input_response",
    source: request.source,
    request_id: request.request_id,
    response_kind: "option",
    option_id: option.id,
    value: option.value,
  };
}

export function buildHumanInputFormSummary(
  request: HumanInputRequest,
  values: Record<string, string | boolean | string[]>,
): string {
  const parts: string[] = [];
  for (const field of request.fields ?? []) {
    const value = values[field.name];
    if (value === undefined || value === "" || (Array.isArray(value) && value.length === 0)) continue;
    const formatted = Array.isArray(value) ? value.join(", ") : typeof value === "boolean" ? (value ? "yes" : "no") : String(value);
    parts.push(`${field.label}: ${formatted}`);
  }
  return parts.join("; ");
}

export function buildHumanInputFormSubmissionValue(
  request: HumanInputRequest,
  values: Record<string, string | boolean | string[]>,
): string {
  const record: Record<string, string | boolean | string[]> = {};
  for (const field of request.fields ?? []) {
    const value = values[field.name];
    if (value === undefined || value === "" || (Array.isArray(value) && value.length === 0)) continue;
    record[field.name] = value;
  }
  return `${buildHumanInputFormSummary(request, values)} [values: ${JSON.stringify(record)}]`;
}

export function createHumanInputTextResponse(
  request: HumanInputRequest,
  value: string,
): HumanInputResponse {
  return {
    version: 1,
    kind: "human_input_response",
    source: request.source,
    request_id: request.request_id,
    response_kind: "text",
    value: value.trim(),
  };
}

export function buildHumanInputResponseText(request: HumanInputRequest, value: string): string {
  return `For your clarification "${request.question}", my answer is: ${value.trim()}`;
}

export function parseHumanInputResponse(value: unknown): HumanInputResponse | null {
  if (!isRecord(value)) return null;
  if (value.version !== 1 || value.kind !== "human_input_response") return null;
  if (!nonEmpty(value.source) || !nonEmpty(value.request_id) || !nonEmpty(value.value)) return null;
  if (value.response_kind === "option") {
    if (!nonEmpty(value.option_id)) return null;
    return {
      version: 1,
      kind: "human_input_response",
      source: value.source,
      request_id: value.request_id,
      response_kind: "option",
      option_id: value.option_id,
      value: value.value,
    };
  }
  if (value.response_kind === "text") {
    return {
      version: 1,
      kind: "human_input_response",
      source: value.source,
      request_id: value.request_id,
      response_kind: "text",
      value: value.value,
    };
  }
  return null;
}

File: backend/agent_runtime/human_input.py
"""DeerFlow-aligned human_input request/response helpers.

Official DeerFlow (upstream) emits ToolMessage.artifact.human_input and accepts
HumanMessage.additional_kwargs.human_input_response. Installed harness 2.1 still
formats ask_clarification as markdown only; we rebuild the native request shape
from tool-call args so the workbench UI can match Human Input Card without a
parallel clarification middleware.
"""

from __future__ import annotations

import re
from typing import Any

# Harness formats options as "1. …" / "  2) …" lines inside the tool result markdown.
_NUMBERED_OPTION_RE = re.compile(
    r"^\s*(?:\d+[\.\、\)]\s+|[-*]\s+)(.+?)\s*$",
    re.MULTILINE,
)
_CONTEXT_EMOJI_RE = re.compile(r"^[🧩❓ℹ️]\s*")


def _non_empty_str(value: Any) -> str | None:
    if not isinstance(value, str):
        return None
    text = value.strip()
    return text or None


def _normalize_options(raw: Any) -> list[dict[str, str]]:
    if isinstance(raw, str):
        try:
            import json

            raw = json.loads(raw)
        except (TypeError, ValueError):
            raw = [raw]
    if raw is None:
        return []
    if not isinstance(raw, list):
        raw = [raw]
    options: list[dict[str, str]] = []
    seen: set[str] = set()
    for idx, item in enumerate(raw, start=1):
        if isinstance(item, dict):
            label = _non_empty_str(item.get("label") or item.get("value") or item.get("id"))
            value = _non_empty_str(item.get("value") or item.get("label") or item.get("id"))
            option_id = _non_empty_str(item.get("id")) or f"opt-{idx}"
        else:
            label = _non_empty_str(item)
            value = label
            option_id = f"opt-{idx}"
        if not label or not value or option_id in seen or value in {o["value"] for o in options}:
            continue
        seen.add(option_id)
        options.append({"id": option_id, "label": label, "value": value})
    return options


def extract_options_from_markdown(text: str | None) -> list[dict[str, str]]:
    """Pull numbered / bulleted choices out of harness markdown result text."""
    if not text:
        return []
    labels: list[str] = []
    for match in _NUMBERED_OPTION_RE.finditer(text):
        label = match.group(1).strip()
        if label:
            labels.append(label)
    return _normalize_options(labels)


def split_clarification_markdown(text: str | None) -> tuple[str | None, str, list[dict[str, str]]]:
    """Split harness markdown into (context, question, options).

    Typical shape::
        🧩 context paragraph…

        第 1 个问题：question text…
        1. option A
        2. option B
    """
    raw = (text or "").strip()
    if not raw:
        return None, "需要你补充一些信息才能继续", []

    options = extract_options_from_markdown(raw)
    body = raw
    if options:
        # Drop trailing numbered list so question/context stay clean.
        lines = raw.splitlines()
        kept: list[str] = []
        for line in lines:
            if _NUMBERED_OPTION_RE.match(line):
                break
            kept.append(line)
        body = "\n".join(kept).strip()

    body = _CONTEXT_EMOJI_RE.sub("", body).strip()
    context: str | None = None
    question = body or raw

    # Prefer an explicit "第 N 个问题：" split when present.
    q_match = re.search(r"(第\s*\d+\s*个问题[:：]\s*)", body)
    if q_match:
        before = body[: q_match.start()].strip()
        after = body[q_match.end() :].strip()
        if before:
            context = before
        if after:
            question = after
    else:
        parts = [p.strip() for p in re.split(r"\n\s*\n", body) if p.strip()]
        if len(parts) >= 2:
            context = parts[0]
            question = "\n\n".join(parts[1:])

    question = re.sub(r"^第\s*\d+\s*个问题[:：]\s*", "", question).strip() or question
    return context, question, options


def extract_human_input_from_tool_result(payload: dict[str, Any] | None) -> dict[str, Any] | None:
    """Prefer DeerFlow-native ToolMessage.artifact.human_input when present."""
    if not isinstance(payload, dict):
        return None
    artifact = payload.get("artifact")
    if isinstance(artifact, dict):
        human_input = artifact.get("human_input")
        if isinstance(human_input, dict) and human_input.get("kind") == "human_input_request":
            # Keep legacy aliases used by older UI/SSE consumers.
            out = dict(human_input)
            call_id = out.get("tool_call_id") or payload.get("call_id")
            if call_id and "call_id" not in out:
                out["call_id"] = call_id
            if call_id and "tool_call_id" not in out:
                out["tool_call_id"] = call_id
            return out
    return None


def build_human_input_request(
    *,
    args: dict[str, Any] | None,
    call_id: str | None,
    formatted_result: str | None = None,
) -> dict[str, Any]:
    """Build a DeerFlow-native human_input_request from ask_clarification args."""
    args = dict(args or {})
    md_context, md_question, md_options = split_clarification_markdown(formatted_result)
    question = (
        _non_empty_str(args.get("question"))
        or _non_empty_str(md_question)
        or "需要你补充一些信息才能继续"
    )
    # If the model stuffed the whole markdown blob into `question`, re-split it.
    if "\n" in question and (
        extract_options_from_markdown(question) or re.search(r"第\s*\d+\s*个问题", question)
    ):
        split_ctx, split_q, split_opts = split_clarification_markdown(question)
        question = split_q or question
        if not md_context:
            md_context = split_ctx
        if not md_options:
            md_options = split_opts

    call = _non_empty_str(call_id) or "unknown"
    request_id = f"clarification:{call}"
    options = _normalize_options(args.get("options")) or md_options
    raw_fields = args.get("fields")
    fields: list[dict[str, Any]] = []
    if isinstance(raw_fields, list):
        for entry in raw_fields:
            if not isinstance(entry, dict):
                continue
            name = _non_empty_str(entry.get("name"))
            label = _non_empty_str(entry.get("label")) or name
            if not name or not label:
                continue
            field_type = str(entry.get("type") or "text")
            fields.append(
                {
                    "name": name,
                    "label": label,
                    "type": field_type,
                    "required": bool(entry.get("required")),
                    **(
                        {"placeholder": str(entry["placeholder"])}
                        if entry.get("placeholder") is not None
                        else {}
                    ),
                    **(
                        {"options": _normalize_options(entry.get("options"))}
                        if entry.get("options") is not None
                        else {}
                    ),
                }
            )

    if fields:
        version, input_mode = 2, "form"
    elif options:
        version, input_mode = 1, "choice_with_other"
    else:
        version, input_mode = 1, "free_text"

    payload: dict[str, Any] = {
        "version": version,
        "kind": "human_input_request",
        "source": "ask_clarification",
        "request_id": request_id,
        "tool_call_id": call,
        "question": question,
        "input_mode": input_mode,
    }
    clarification_type = _non_empty_str(args.get("clarification_type"))
    if clarification_type:
        payload["clarification_type"] = clarification_type
    context = args.get("context")
    if context is None or isinstance(context, str):
        if _non_empty_str(context):
            payload["context"] = str(context).strip()
        elif md_context:
            payload["context"] = md_context
        else:
            payload["context"] = None
    if input_mode == "form":
        payload["fields"] = fields
    elif options:
        payload["options"] = options
    # Keep legacy aliases for existing SSE consumers / activity summary.
    payload["call_id"] = call
    return payload


def normalize_human_input_response(raw: Any) -> dict[str, Any] | None:
    """Validate outbound human_input_response metadata (DeerFlow-native v1)."""
    if not isinstance(raw, dict):
        return None
    if raw.get("version") != 1 or raw.get("kind") != "human_input_response":
        return None
    source = _non_empty_str(raw.get("source")) or "ask_clarification"
    request_id = _non_empty_str(raw.get("request_id"))
    value = _non_empty_str(raw.get("value"))
    if not request_id or not value:
        return None
    response_kind = raw.get("response_kind")
    if response_kind == "text":
        return {
            "version": 1,
            "kind": "human_input_response",
            "source": source,
            "request_id": request_id,
            "response_kind": "text",
            "value": value,
        }
    if response_kind == "option":
        option_id = _non_empty_str(raw.get("option_id"))
        if not option_id:
            return None
        return {
            "version": 1,
            "kind": "human_input_response",
            "source": source,
            "request_id": request_id,
            "response_kind": "option",
            "option_id": option_id,
            "value": value,
        }
    return None


def build_response_message_text(question: str, value: str) -> str:
    """Readable bridge text (matches DeerFlow frontend template intent)."""
    q = question.strip() or "your clarification"
    return f'For your clarification "{q}", my answer is: {value.strip()}'
